app-backend/api/src/main/scala/uploads/Routes.scala

```diff
@@ -51,240 +56,300 @@ trait UploadRoutes
-            UploadDao.query
-              .filter(uploadId)
-              .selectOption
-              .transact(xa)
-              .unsafeToFuture
+  def getUpload(uploadId: UUID): Route =
+    authenticate { user =>
+      authorizeScope(ScopedAction(Domain.Uploads, Action.Read, None), user) {
+        authorizeAsync {
+          UploadDao.query
+            .ownedByOrSuperUser(user, uploadId)
+            .exists
+            .transact(xa)
+            .unsafeToFuture
+        } {
+          rejectEmptyResponse {
+            complete {
+              UploadDao.query
+                .filter(uploadId)
+                .selectOption
+                .transact(xa)
+                .unsafeToFuture
+            }
           }
         }
       }
     }
-  }
 
-  def createUpload: Route = authenticate { user =>
-    val userBytesUploaded =
-      UploadDao.getUserBytesUploaded(user).transact(xa).unsafeToFuture()
-    authorizeScopeLimit(
-      userBytesUploaded,
-      Domain.Uploads,
-      Action.Create,
-      user
-    ) {
-      entity(as[Upload.Create]) { newUpload =>
-        logger.debug(
-          s"newUpload: ${newUpload.uploadType}, ${newUpload.source}, ${newUpload.fileType}"
-        )
-        val uploadToInsert =
-          (newUpload.uploadType, newUpload.source, newUpload.fileType) match {
-            case (UploadType.S3, _, FileType.NonSpatial) => {
-              if (newUpload.files.nonEmpty) newUpload
-              else
-                throw new IllegalStateException(
-                  "S3 upload must specify a source if no files are specified"
-                )
-            }
-            case (UploadType.S3, Some(source), _) => {
-              if (newUpload.files.nonEmpty) newUpload
-              else {
-                val files = listAllowedFilesInS3Source(source)
-                if (files.nonEmpty) newUpload.copy(files = files)
+  def createUpload: Route =
+    authenticate { user =>
+      // check if user has already hit the storage quota
+      val userBytesUploaded =
+        UploadDao.getUserBytesUploaded(user).transact(xa).unsafeToFuture()
+      authorizeScopeLimit(
+        userBytesUploaded,
+        Domain.Uploads,
+        Action.Create,
+        user
+      ) {
+        entity(as[Upload.Create]) { newUpload =>
+          logger.debug(
+            s"newUpload: ${newUpload.uploadType}, ${newUpload.source}, ${newUpload.fileType}"
+          )
+          // check the bytes uploaded to this S3 location
+          val bytesUploaded = Upload.getBytesUploaded(
+            s3.client,
+            dataBucket,
+            newUpload.files,
+            newUpload.uploadStatus,
+            newUpload.uploadType
+          )
+          // check the file size in bytes submitted by this request
+          // the potential size should be the max of
+          // the fileSize and the S3 file size check
+          val potentialNewBytes =
+            newUpload.fileSizeBytes.getOrElse(0.toLong).max(bytesUploaded)
```

Comment:
My recollection is that we avoided storing this on the upload object initially because a medium-sophistication user could just lie. Since we have to get the size in bytes from s3 anyway, what value does storing the value on the upload provide?

Reply:
I want to have a call with you about this. I want to go through the general upload workflow and make sure we are on the same page. We can also discuss the suggestion you made below.